[video] Fix Versions Art Changes

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Change the way movie / versions artwork is stored and retrieved to provide an intuitive and consistent ux.
As a nice side-effect, this restored "thumbnails" as default art for extras, overridable by user.
It should "just work" but here are additional details.

From the user's point of view, the art presented for the movie is the art of the default version (except for version-aware dialogs/nav). Here what it means for different contexts:

1/ Anywhere in Kodi, except 'Manage versions', 'Choose version', 'versions as folder'
  - The art comes from the default version of the movie. Changing art changes the art of the default version and is visible everywhere that's not version-aware.

2/ Manage versions (version-aware)
  - Changing the art of the default version also changes the art visible for the movie outside of Manage versions
  - Changes of the art of a non-default version is visible only in version-aware contexts
  - Changing the default version switches art everywhere that's not version-aware to the art of the new default version

3/ Choose version, Versions as folder (version-aware)
  - Show the individual versions art, default version or not.

Also fixed a few places where changing art was not refreshing the calling dialogs (probably a few left to do).
fanart will require another look, it's handled differently and Info dialog behaves differently as well.

**Technical details**

When reading or setting art for a movie, the code that is not aware of version art (all code except for the manage versions dialog) will be redirected to automatically use the default version art instead (stored with 'videoversion' media_type instead of 'movie').

There is no change for the retrieval/setting of art with a media_type other 'movie'.

The media_type 'movie' is not used in the art table anymore. All existing 'movie' art is migrated automatically to 'videoversion' art during the database upgrade. Manual user changes of the default version art have priority over art of matching 'movie' art.

This may seem like a radical change but it follows the direction of evolving to an asset table that has basic attributes such as art. Conceptually, a record of videoversion is pretty close to a future "asset". In v22 we'll reintroduce movie groups, that will have their art combined with the art of the assets (much like movie sets).

**Some notes:**
- database upgrade code: the accomplished action is an UPDATE. However UPDATE with additional tables has different syntaxes with SQLite and mySql/MariaDB. An INSERT followed by DELETE is more efficient than many single-row UPDATE statements. The DELETE would have been needed anyway after the individual updates to cleanup the 'movie' art that already had an equivalent as 'videoversion'.
- it would be possible to avoid the artwork redirection from movie to videoversion for callers that already have a videotag with the required info. The saving would be rather small though, I made it so that the conversion is efficient in this PR. The number of queries would not change (1) and it's only a join supported by an index that would be saved. Maybe as a future PR if there is a need.
- Took care of some dialog artwork refresh, there may be some left to do. It wasn't the core objective of the PR.
- Something remains to be done for fanart, though some problems already existed and some combinations already work.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

There was an overlap between movie art and default version art which lead to inconsistencies and big UI changes are not an option at this point in the v21 cycle.

There are alternative ways to the proposal but I didn't find one that also didn't create other inconsistencies or eventually leaked the art between versions.

The alternative of cutting version art didn't seem desirable.

v22 will re-add the concept of "movie group" art that will be merged with individual movies/versions art for display, but with proper UI to manage.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

With new database and database migrated from before versions, on sqlite and maria db.
Created versions, changed art of movie, art of versions, changed default version, ...

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
More intuitive and predictable behavior of movie and version art.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[videodb] Store/retrieve art of media_type 'movie' as 'videoversion' + default version's id.

CrystalP committed:
[video] Fix art not refreshing when changing version art or default version

CrystalP commented on xbmc/video/dialogs/GUIDialogVideoManager.cpp:
```diff
@@ -330,6 +330,15 @@ void CGUIDialogVideoManager::ChooseArt()
+
+  // notify all windows to update the file item
+  CGUIMessage msg{GUI_MSG_NOTIFY_ALL,        0,           0, GUI_MSG_UPDATE_ITEM,
+                  GUI_MSG_FLAG_FORCE_UPDATE, m_videoAsset};
```
This formatting required by clang / Jenkins. There has to be a way to tune this?

ksooo replied:
No idea.

garbear replied:
In clang-format we trust. But it subjectively flunks some cases. Not hard to fix:

```
// clang-format off
...
// clang-format on`
```

CrystalP replied:
yes thanks, as last resort.

CrystalP commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -183,6 +184,10 @@ void CGUIDialogVideoManagerVersions::SetDefaultVideoVersion(const CFileItem& ver
+  thumbLoader.LoadItem(m_videoAsset.get());
+
   // notify all windows to update the file item
   CGUIMessage msg{GUI_MSG_NOTIFY_ALL,        0,           0, GUI_MSG_UPDATE_ITEM,
```
This formatting required by clang / Jenkins. There has to be a way to tune this?

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4864,10 +4864,23 @@ void CVideoDatabase::SetVideoSettings(int idFile, const CVideoSettings &setting)
+  }
+  else
+  {
+    for (const auto& i : art)
+      SetArtForItem(mediaId, mediaType, i.first, i.second);
```
You can simplify this by calling `void SetArtForItem(int mediaId, const MediaType &mediaType, const std::map<std::string, std::string> &art);` - no need to iterate yourself.

CrystalP replied:
That would cause infinite recursion. This block is the original code. I added the block above to translate the media_id / media_type parameters and call the original code through 1 level of recursion, it will never go beyond that.
Even that recursion could be avoided, but I didn't find a way that would make extra copies along the way.

ksooo replied:
Alright.

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4864,10 +4864,23 @@ void CVideoDatabase::SetVideoSettings(int idFile, const CVideoSettings &setting)
+  {
+    // Take the hit of looking up the default version of the movie only once, instead of once per
+    // art piece.
+    CFileItem item{};
```
No need for explicit init. `CFileItem item;` would be enough.

CrystalP replied:
yes

ksooo commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4951,47 +4953,34 @@ bool CVideoDatabase::GetArtForAsset(int assetId,
-        // insert if not yet present
-        art.insert(make_pair(m_pDS2->fv(1).get_asString(), m_pDS2->fv(2).get_asString()));
-      }
+      art.insert(make_pair(m_pDS2->fv(0).get_asString(), m_pDS2->fv(1).get_asString()));
```
While you are at it, could you please add the `std` namespace to the `make_pair`call?

CrystalP replied:
sure.

ksooo commented on xbmc/video/VideoDatabase.h:
```diff
@@ -945,11 +945,7 @@ class CVideoDatabase : public CDatabase
-                      int ownerId,
-                      const MediaType& mediaType,
-                      std::map<std::string, std::string>& art);
+  std::string GetArtForItem(int mediaId, const MediaType& mediaType, const std::string& artType);
```
Could this method be `const`? I don't think it changes state?

CrystalP replied:
Oddly it could, but through a bit of unvoluntary trickery I believe. m_pDS2 is modified in the end.
I didn't touch that function because I haven't seen it called for movies and only "TV recently added" but it would be safer to change it now as well.

CrystalP replied:
Actually there also are RemoveArtForItem() function to take care of, as well as HasArtForItem()

ksooo reviewed (approved):
I trust on your sql skills. The c++ code looks good to me, except some minor nitpicks. It seems you moved parts of my previous approach down one layer, to the database?

ksooo commented:
Jenkins build this please

CrystalP commented:
Win Jenkins failures were likely for the same reason as user here https://forum.kodi.tv/showthread.php?tid=375688
Now that the file is on the mirrors build should work again.

CrystalP commented:
> It seems you moved parts of my previous approach down one layer, to the database?
Yes in a way. There is no "clean" way to implement this because it's non-standard to have file items of type movie that use art that is videoversion. Deeper changes would be required and that feels like v22 material.

Addressing my own notes, HasArtForItem() is only used for movie sets. I added a todo inside.
RemoveArtForItem doesn't have users in core Kodi. It's exposed by json-rpc however, so I modified the functions.
I'm not very json-rpc knowledgeable and don't know how to test it though. No change for media_types other than 'movie'.

I'd be more comfortable with a look at the SQL by @da-anda or @the-black-eagle, especially for the data migration part.

ksooo commented:
Jenkins build this please

ksooo commented:
Jenkins build this please

CrystalP commented:
I'm rethinking this, there may be a less risky way to do this and to postpone some changes into v22.
